{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "00670c4a-cc62-4b14-95ff-f6b25e36412e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4.0.0\n"
     ]
    }
   ],
   "source": [
    "#Import the dependencies for this workflow\n",
    "import sys\n",
    "import os\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import nupack\n",
    "import pickle\n",
    "from tkinter import Tk\n",
    "from tkinter.filedialog import askdirectory\n",
    "from nupack import * # This star means import all\n",
    "\n",
    "print(nupack.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cf35b60d-f60e-443c-9d5b-491be99d96f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Toehold Switch VISTA Design Code \n",
    "\n",
    "# Step 1: Create toehold switches tiled across source sequence inputs, at specified temperature, with output reporter sequence\n",
    "# VISTA desgin code is based on tsgen2 toehold switch architechure. Nucleotides 7-18 of the ascending stem, ...\n",
    "# the loop, and nucleotides of the descending stem, including the start codon AUG are conserved. The sequence for this conserved element is: GUUAUAGUUAUGAACAGAGGAGACAUAACAUGAAC\n",
    "# For these desings, a 36-nt toehold is used, with an unwinding of 6-bp in the stem. \n",
    "\n",
    "class toehold_switch_design_object():\n",
    "\n",
    "    # The initialization step takes in a file called 'source_seq_list.csv' and saves info uniquely to this instance of the toehold design class\n",
    "    # Parameters will change depending on the design specified in 'source_seq_list.csv'\n",
    "    # The column names in 'source_seq_list.csv' must remain constant. \n",
    "    def __init__(self,source_seq_list,ecoli_codon_usage_table,target_type,num_designs):\n",
    "        self.path = askdirectory(title='Select Folder')\n",
    "        os.chdir(self.path)\n",
    "        input_table = pd.read_csv(source_seq_list)\n",
    "        self.mRNA_names0 = input_table[['Gene Name']]\n",
    "        self.mRNA_sequences0 = input_table[['Gene Sequence']]\n",
    "        self.temp = input_table[['Temperature']]\n",
    "        self.output_name = input_table[['Output Name']]\n",
    "        self.output_sequence = input_table[['Output Sequence']]\n",
    "        self.hairpin_seq_top = 'GUUAUAGUUAUGAACAGAGGAGACAUAACAUGAAC' #top part of the toehold switch stem/loop, including RBS \n",
    "        self.hairpin_suffix = 'AACCUGGCGGCAGCGCAAAAG' #linker sequence between toehold switch & reporter output \n",
    "        mRNA_names = {}; mRNA_sequences = {}; mRNA_temps = [];\n",
    "        self.num_designs = num_designs\n",
    "\n",
    "        self.design_type = 'Toehold_VISTA';\n",
    "        os.mkdir(self.design_type);\n",
    "        self.target_type = target_type;\n",
    "        self.codon_usage_df = pd.read_csv(ecoli_codon_usage_table); self.codon_usage_dict = dict(zip(self.codon_usage_df[\"Codon\"], self.codon_usage_df[\"Fraction\"]))\n",
    "\n",
    "    def reverse_complement_seq(self,nucleotide_sequence):\n",
    "        \"\"\"\n",
    "        Calculate the reverse complement of an RNA sequence.\n",
    "        Args: rna_sequence (str):  A nucleotide sequence consisting of 'A', 'T'/'U', 'C', 'G'.\n",
    "        Returns: str: The reverse complement of the RNA sequence.\n",
    "        \"\"\"\n",
    "        nucleotide_sequence = nucleotide_sequence.upper()\n",
    "        # Define the complement mapping for RNA\n",
    "        complement = {\"A\": \"T\", \"T\": \"A\", \"C\": \"G\", \"G\": \"C\"}\n",
    "        # Reverse the sequence and create its complement\n",
    "        reverse_sequence = nucleotide_sequence[::-1]\n",
    "        reverse_complement = [complement[base] for base in reverse_sequence]\n",
    "        # Join the complement bases to form the reverse complement sequence\n",
    "        reverse_complement_sequence = \"\".join(reverse_complement)\n",
    "\n",
    "        return reverse_complement_sequence\n",
    "\n",
    "    def dna_to_rna(self,dna_sequence):\n",
    "        dna_sequence = dna_sequence.upper() \n",
    "        return dna_sequence.replace('T', 'U')\n",
    "\n",
    "    def rna_to_dna(self,dna_sequence):\n",
    "        dna_sequence = dna_sequence.upper() \n",
    "        return dna_sequence.replace('U', 'T')\n",
    "\n",
    "    def nt_to_aa(self,seq):\n",
    "        \"\"\"Convert nucleotide sequence to amino acids.\"\"\"\n",
    "        codon_table = {\n",
    "            'UUU':'F', 'UUC':'F', 'UUA':'L', 'UUG':'L','UCU':'S', 'UCC':'S', 'UCA':'S', 'UCG':'S',\n",
    "            'UAU':'Y', 'UAC':'Y', 'UAA':'*', 'UAG':'*','UGU':'C', 'UGC':'C', 'UGA':'*', 'UGG':'W',\n",
    "            'CUU':'L', 'CUC':'L', 'CUA':'L', 'CUG':'L','CCU':'P', 'CCC':'P', 'CCA':'P', 'CCG':'P',\n",
    "            'CAU':'H', 'CAC':'H', 'CAA':'Q', 'CAG':'Q','CGU':'R', 'CGC':'R', 'CGA':'R', 'CGG':'R',\n",
    "            'AUU':'I', 'AUC':'I', 'AUA':'I', 'AUG':'M','ACU':'T', 'ACC':'T', 'ACA':'T', 'ACG':'T',\n",
    "            'AAU':'N', 'AAC':'N', 'AAA':'K', 'AAG':'K','AGU':'S', 'AGC':'S', 'AGA':'R', 'AGG':'R',\n",
    "            'GUU':'V', 'GUC':'V', 'GUA':'V', 'GUG':'V','GCU':'A', 'GCC':'A', 'GCA':'A', 'GCG':'A',\n",
    "            'GAU':'D', 'GAC':'D', 'GAA':'E', 'GAG':'E','GGU':'G', 'GGC':'G', 'GGA':'G', 'GGG':'G'\n",
    "        }\n",
    "        aa_seq = \"\"\n",
    "        for i in range(0, len(seq)-2, 3):\n",
    "            codon = seq[i:i+3]\n",
    "            aa_seq += codon_table.get(codon, 'X')  # 'X' for unknown codons\n",
    "        return aa_seq\n",
    "\n",
    "    def compute_sensor_parameters(self,sensor_sequence_r):\n",
    "        \"\"\"\n",
    "        Calculates minimum free energy (MFE) and ensemble defects of various regions of the toehold switch sensor. \n",
    "        Args: sensor_sequence_r (str):  An RNA sequence consisting of 'A', 'U', 'C', 'G'.\n",
    "        Returns: appended columns to output_table: A dictionary containing computed thermodynamic values for toehold switch sensor. \n",
    "        \"\"\"\n",
    "        #Define the ideal structure for various regions of the toehold switch sensor; IMPORTANT: these are defined based on the TSgen2 Toehold Switch Structure, where the toehold is 30nt, and the target invades 6nt into the stem. \n",
    "        AS = sensor_sequence_r[30:48]; AS_struc = '.' * len(AS); #ascending stem\n",
    "        DS = sensor_sequence_r[58:77]; DS_struc = '.' * len(DS); #descending stem\n",
    "        RBS_GFP = sensor_sequence_r[48:116]; RBS_GFP_struc = 'U%s D%s U%s U%s'%(len(self.hairpin_seq_top[12:-3]),(self.len_d+self.len_a),(self.len_b+self.len_a),(self.hairpin_suffix_len+len(self.output_seq30[:9])));\n",
    "        RBS_Linker = sensor_sequence_r[48:107]; RBS_Linker_struc = 'U%s D%s U%s U%s'%(len(self.hairpin_seq_top[12:-3]),(self.len_d+self.len_a),(self.len_b+self.len_a),self.hairpin_suffix_len);\n",
    "        Switch_Off = sensor_sequence_r[:77]; Switch_Off_struc = 'U%s D%s D3 (U3 D6 U11 U3)'%(self.len_c,(self.len_a+self.len_b));\n",
    "        Switch_Off_GFP = sensor_sequence_r[:116]; Switch_Off_GFP_struc = 'U%s D%s D3 (U3 D6 U11 U3) U%s U%s'%(self.len_c,(self.len_a+self.len_b),(self.len_d+self.len_a+3),(self.hairpin_suffix_len+len(self.output_seq30[:9])));\n",
    "        Switch_Off_NoTH = sensor_sequence_r[30:116]; Switch_Off_NoTH_struc = 'D%s D3 (U3 D6 U11 U3) U%s U%s'%((self.len_a+self.len_b),(self.len_d+self.len_a+3),(self.hairpin_suffix_len+len(self.output_seq30[:9])));\n",
    "        Hairpin = sensor_sequence_r[30:77]; Hairpin_struc = 'D%s D3 (U3 D6 U11 U3)'%(self.len_a+self.len_b);\n",
    "        Linker_GFP = sensor_sequence_r[77:116]; TH_Linker_struc = 'U%s + U%s'%(self.len_c,(self.len_a+self.len_b+self.len_d+self.hairpin_suffix_len+len(self.output_seq30[:9])));TH = sensor_sequence_r[:30];\n",
    "        Switch_On = sensor_sequence_r[:76]; Switch_On_struc = 'D%s + U%s'%((self.len_c+self.len_a+self.len_b),(len(self.hairpin_seq_top)+self.len_a+self.len_b))\n",
    "        Switch_On_GFP_struc = 'D%s + U%s D6 U11 U3 D%s U%s U%s'%((self.toehold_len+self.len_a+self.len_b),(self.len_a+self.len_b), (3+self.len_a), (self.len_b+self.len_d), (self.hairpin_suffix_len+len(self.output_seq30[:9])))\n",
    "        \n",
    "        #Calcualte minimum free energy (MFE) parameters\n",
    "        mfe_structureAS = mfe(strands=[AS], model=self.my_model); self.output_table[f\"Ascending Stem MFE\"].append(mfe_structureAS[0].energy);\n",
    "        mfe_structureDS = mfe(strands=[DS], model=self.my_model); self.output_table[f\"Descending Stem MFE\"].append(mfe_structureDS[0].energy);\n",
    "        mfe_structureRBS_GFP = mfe(strands=[RBS_GFP], model=self.my_model); self.output_table[f\"RBS-GFP Stem MFE\"].append(mfe_structureRBS_GFP[0].energy);\n",
    "        mfe_structureRBS_Linker = mfe(strands=[RBS_Linker], model=self.my_model); self.output_table[f\"RBS-Linker MFE\"].append(mfe_structureRBS_Linker[0].energy);\n",
    "        mfe_structureSwitch_Off = mfe(strands=[Switch_Off], model=self.my_model); self.output_table[f\"Switch Off MFE\"].append(mfe_structureSwitch_Off[0].energy);\n",
    "        mfe_structureSwitch_Off_GFP = mfe(strands=[Switch_Off_GFP], model=self.my_model); self.output_table[f\"Switch Off GFP MFE\"].append(mfe_structureSwitch_Off_GFP[0].energy);\n",
    "        mfe_structureSwitch_Off_NoTH = mfe(strands=[Switch_Off_NoTH], model=self.my_model); self.output_table[f\"Switch Off NoTH MFE\"].append(mfe_structureSwitch_Off_NoTH[0].energy);\n",
    "        mfe_structureHairpin = mfe(strands=[Hairpin], model=self.my_model); self.output_table[f\"Hairpin MFE\"].append(mfe_structureHairpin[0].energy);\n",
    "        mfe_structureTH_Linker = mfe(strands=[TH, Linker_GFP], model=self.my_model); self.output_table[f\"TH+Linker/GFP MFE\"].append(mfe_structureTH_Linker[0].energy);\n",
    "        mfe_structureSwitch_On = mfe(strands=[self.target_sequence,Switch_Off], model=self.my_model); self.output_table[f\"Switch ON MFE\"].append(mfe_structureSwitch_On[0].energy);\n",
    "        mfe_structureSwitch_On_GFP = mfe(strands=[self.target_sequence,Switch_Off_GFP], model=self.my_model); self.output_table[f\"Switch ON-GFP MFE\"].append(mfe_structureSwitch_On_GFP[0].energy);\n",
    "    \n",
    "        #Calcualte ideal ensemble defect, or the defect from the specified structure of the toehold switch.\n",
    "        ensemble_defect_idealAS = defect(strands=[AS], structure= AS_struc, model=self.my_model); self.output_table[f\"Ascending Stem IED\"].append(ensemble_defect_idealAS)\n",
    "        ensemble_defect_idealDS = defect(strands=[DS], structure= DS_struc, model=self.my_model); self.output_table[f\"Descending Stem IED\"].append(ensemble_defect_idealDS)\n",
    "        ensemble_defect_idealRBS_GFP = defect(strands=[RBS_GFP], structure=RBS_GFP_struc, model=self.my_model); self.output_table[f\"RBS-GFP Stem IED\"].append(ensemble_defect_idealRBS_GFP)\n",
    "        ensemble_defect_idealRBS_Linker = defect(strands=[RBS_Linker], structure=RBS_Linker_struc, model=self.my_model); self.output_table[f\"RBS-Linker IED\"].append(ensemble_defect_idealRBS_Linker)\n",
    "        ensemble_defect_idealSwitch_Off = defect(strands=[Switch_Off], structure=Switch_Off_struc, model=self.my_model); self.output_table[f\"Switch Off IED\"].append(ensemble_defect_idealSwitch_Off)\n",
    "        ensemble_defect_idealSwitch_Off_GFP = defect(strands=[Switch_Off_GFP], structure=Switch_Off_GFP_struc, model=self.my_model); self.output_table[f\"Switch Off GFP IED\"].append(ensemble_defect_idealSwitch_Off_GFP)\n",
    "        ensemble_defect_idealSwitch_Off_NoTH = defect(strands=[Switch_Off_NoTH], structure=Switch_Off_NoTH_struc, model=self.my_model); self.output_table[f\"Switch Off NoTH IED\"].append(ensemble_defect_idealSwitch_Off_NoTH)\n",
    "        ensemble_defect_idealHairpin = defect(strands=[Hairpin], structure=Hairpin_struc, model=self.my_model); self.output_table[f\"Hairpin IED\"].append(ensemble_defect_idealHairpin)\n",
    "        ensemble_defect_idealTH_Linker = defect(strands=[TH,Linker_GFP], structure=TH_Linker_struc, model=self.my_model); self.output_table[f\"TH+Linker/GFP IED\"].append(ensemble_defect_idealTH_Linker)\n",
    "        ensemble_defect_idealSwitch_On = defect(strands=[self.target_sequence,Switch_Off], structure=Switch_On_struc, model=self.my_model); self.output_table[f\"Switch ON IED\"].append(ensemble_defect_idealSwitch_On)\n",
    "        ensemble_defect_idealSwitch_On_GFP = defect(strands=[self.target_sequence,Switch_Off_GFP], structure=Switch_On_GFP_struc, model=self.my_model); self.output_table[f\"Switch ON-GFP IED\"].append(ensemble_defect_idealSwitch_On_GFP)\n",
    "        \n",
    "        #Calculate native ensemble defect, or the defect from the MFE proxy structure calculated earlier in this fuction. \n",
    "        ensemble_defect_nativeAS = defect(strands=[AS], structure=str(mfe_structureAS[0].structure), model=self.my_model); self.output_table[f\"Ascending Stem NED\"].append(ensemble_defect_nativeAS)\n",
    "        ensemble_defect_nativeDS = defect(strands=[DS], structure=str(mfe_structureDS[0].structure), model=self.my_model); self.output_table[f\"Descending Stem NED\"].append(ensemble_defect_nativeDS)\n",
    "        ensemble_defect_nativeRBS_GFP = defect(strands=[RBS_GFP], structure=str(mfe_structureRBS_GFP[0].structure), model=self.my_model); self.output_table[f\"RBS-GFP Stem NED\"].append(ensemble_defect_nativeRBS_GFP)\n",
    "        ensemble_defect_nativeRBS_Linker = defect(strands=[RBS_Linker], structure=str(mfe_structureRBS_Linker[0].structure), model=self.my_model); self.output_table[f\"RBS-Linker NED\"].append(ensemble_defect_nativeRBS_Linker)\n",
    "        ensemble_defect_nativeSwitch_Off = defect(strands=[Switch_Off], structure=str(mfe_structureSwitch_Off[0].structure), model=self.my_model); self.output_table[f\"Switch Off NED\"].append(ensemble_defect_nativeSwitch_Off)\n",
    "        ensemble_defect_nativeSwitch_Off_GFP = defect(strands=[Switch_Off_GFP], structure=str(mfe_structureSwitch_Off_GFP[0].structure), model=self.my_model); self.output_table[f\"Switch Off GFP NED\"].append(ensemble_defect_nativeSwitch_Off_GFP)\n",
    "        ensemble_defect_nativeSwitch_Off_NoTH = defect(strands=[Switch_Off_NoTH], structure=str(mfe_structureSwitch_Off_NoTH[0].structure), model=self.my_model); self.output_table[f\"Switch Off NoTH NED\"].append(ensemble_defect_nativeSwitch_Off_NoTH)\n",
    "        ensemble_defect_nativeHairpin = defect(strands=[Hairpin], structure=str(mfe_structureHairpin[0].structure), model=self.my_model); self.output_table[f\"Hairpin NED\"].append(ensemble_defect_nativeHairpin)\n",
    "        ensemble_defect_nativeTH_Linker = defect(strands=[TH,Linker_GFP], structure=str(mfe_structureTH_Linker[0].structure), model=self.my_model); self.output_table[f\"TH+Linker/GFP NED\"].append(ensemble_defect_nativeTH_Linker)\n",
    "        ensemble_defect_nativeSwitch_On = defect(strands=[self.target_sequence,Switch_Off], structure=str(mfe_structureSwitch_On[0].structure), model=self.my_model); self.output_table[f\"Switch ON NED\"].append(ensemble_defect_nativeSwitch_On)\n",
    "        ensemble_defect_nativeSwitch_On_GFP = defect(strands=[self.target_sequence,Switch_Off_GFP], structure=str(mfe_structureSwitch_On_GFP[0].structure), model=self.my_model); self.output_table[f\"Switch ON-GFP NED\"].append(ensemble_defect_nativeSwitch_On_GFP)\n",
    "\n",
    "    def compute_target_parameters(self,target_sequence):\n",
    "        \"\"\"\n",
    "        Calculates minimum free energy (MFE) and ensemble defects of various regions of the target around each truncated 36nt binding site. \n",
    "        Optimally, the target strand is single stranded, indicating an MFE~0 and an ensemble defect ~0\n",
    "        Args: target_sequence (str):  An RNA sequence consisting of 'A', 'U', 'C', 'G'.\n",
    "        Returns: appended columns to output_table: A dictionary containing computed thermodynamic values for toehold switch target and various predetermined flanking lengths. \n",
    "        \"\"\"\n",
    "        mRNA_seq_RNA = self.dna_to_rna(self.mRNA_seq)\n",
    "        index = mRNA_seq_RNA.find(target_sequence)\n",
    "        for length in self.flanking_lengths:\n",
    "            left_flank = max(0, index - length)\n",
    "            right_flank = min(index + len(self.target_sequence) + length, len(mRNA_seq_RNA))\n",
    "            flanking_sequence = mRNA_seq_RNA[left_flank:index] + self.target_sequence + mRNA_seq_RNA[index+len(self.target_sequence):right_flank]\n",
    "            mfe_structures = mfe(strands=[flanking_sequence], model=self.my_model)\n",
    "            native_structure = mfe_structures[0].structure; ideal_structure = '.' * len(flanking_sequence)\n",
    "            partition_function = pfunc(strands=[flanking_sequence], model=self.my_model); parfunc = partition_function[1]\n",
    "            self.output_table[f\"Trigger (+-{length} nt) MFE\"].append(mfe_structures[0].energy)\n",
    "            self.output_table[f\"Trigger (+-{length} nt) Ensemble Defect Native\"].append(defect(strands=[flanking_sequence], structure=str(native_structure), model=self.my_model))\n",
    "            self.output_table[f\"Trigger (+-{length} nt) Ensemble Defect Ideal\"].append(defect(strands=[flanking_sequence], structure=ideal_structure, model=self.my_model))\n",
    "\n",
    "    def compute_codon_fractions(self, mRNA_seq, target_sequence, codon_usage_table):\n",
    "        mRNA_seq_DNA = mRNA_seq.upper().replace(\"U\", \"T\"); target_seq = target_sequence.replace(\"U\", \"T\")  # Convert RNA sequence to DNA (U -> T); Convert target sequence to DNA (U -> T)\n",
    "        target_start = mRNA_seq_DNA.find(target_seq)  # Find the target sequence in the mRNA sequence\n",
    "        if target_start == -1:\n",
    "            raise ValueError(\"Target sequence not found in mRNA sequence.\")\n",
    "        closest_in_frame_start = target_start - (target_start % 3)  # Adjust to the closest in-frame codon\n",
    "        codons = [mRNA_seq_DNA[i:i+3] for i in range(0, len(mRNA_seq_DNA) - 2, 3)]  # Extract codons\n",
    "        codon_fractions = [self.codon_usage_dict.get(c, 0) for c in codons]  # Get codon fractions\n",
    "        closest_codon_idx = closest_in_frame_start // 3\n",
    "        target_end_idx = closest_codon_idx + (len(target_seq) // 3)\n",
    "        # Get fractions of the first two codons\n",
    "        codon1_fraction = codon_fractions[closest_codon_idx]; codon2_fraction = codon_fractions[closest_codon_idx + 1];  avg_two_codons = (codon1_fraction + codon2_fraction) / 2\n",
    "        # Compute average codon fraction across the target binding region\n",
    "        avg_target_region = sum(codon_fractions[closest_codon_idx:target_end_idx]) / (target_end_idx - closest_codon_idx)\n",
    "        # Compute preceding codon averages\n",
    "        preceding_averages = {}\n",
    "        for n in range(4, 11):\n",
    "            start_idx = max(0, closest_codon_idx - n)\n",
    "            valid_codon_count = closest_codon_idx - start_idx\n",
    "            if valid_codon_count > 0:\n",
    "                avg_value = sum(codon_fractions[start_idx:closest_codon_idx]) / valid_codon_count\n",
    "            else:\n",
    "                # If no preceding codons, use the first codon fraction as the default\n",
    "                avg_value = codon1_fraction\n",
    "            preceding_averages[f\"Preceding {n} codons avg\"] = avg_value\n",
    "            self.output_table[f\"Preceding {n} codons avg\"].append(avg_value)\n",
    "        # Compute following codon averages\n",
    "        following_averages = {}\n",
    "        for n in range(4, 11):\n",
    "            end_idx = min(len(codon_fractions), target_end_idx + n)\n",
    "            valid_codon_count = end_idx - target_end_idx\n",
    "            if valid_codon_count > 0:\n",
    "                avg_value = sum(codon_fractions[target_end_idx:end_idx]) / valid_codon_count\n",
    "            else:\n",
    "                # If no following codons, use the last codon fraction as the default\n",
    "                avg_value = codon_fractions[-1]\n",
    "            following_averages[f\"Following {n} codons avg\"] = avg_value\n",
    "            self.output_table[f\"Following {n} codons avg\"].append(avg_value)\n",
    "        # Append results to the output table\n",
    "        self.output_table[f\"First Codon Fraction\"].append(codon1_fraction)\n",
    "        self.output_table[f\"Second Codon Fraction\"].append(codon2_fraction)\n",
    "        self.output_table[f\"Average First Two codons\"].append(avg_two_codons)\n",
    "        self.output_table[f\"Average Target Fraction\"].append(avg_target_region)\n",
    "    \n",
    "    def rank_new_designs(self, new_designs, target_type, model_params_path=\"all_trained_model_params.pkl\"):\n",
    "        \"\"\"\n",
    "        Rank new designs using all five PLS-DA models (OFF AVG, ON AVG Truncated, ON AVG Full, ON OFF Truncated, ON OFF Full) and combine rankings.\n",
    "        \n",
    "        Parameters:\n",
    "            new_designs (np.array or pd.DataFrame): Array or DataFrame of new designs (shape: n_new_designs x n_features).\n",
    "            model_params_path (str): Path to the saved model parameters. DO NOT CHANGE: always should be all_trained_model_params.pkl\n",
    "            target_type (str): \"Truncated\" or \"Full\" mRNA target; user selects which type to be ranked.\n",
    "        \n",
    "        Returns:\n",
    "            final_ranking (pd.DataFrame): Final ranking of designs with combined scores.\n",
    "            ranked_indices (np.array): Indices of the ranked designs in the original dataset.\n",
    "        \"\"\"\n",
    "        with open(model_params_path, \"rb\") as f:  # Load the saved parameters\n",
    "            model_params = pickle.load(f)\n",
    "    \n",
    "        rankings = {}  # Initialize a dictionary to store probabilities for each model\n",
    "\n",
    "        for model_name, params in model_params.items():  # Compute probabilities for each model\n",
    "            # Get the selected feature indices for this model\n",
    "            selected_feature_indices = params[\"selected_feature_indices\"]\n",
    "    \n",
    "            # Filter new_designs to include only the selected features (columns)\n",
    "            if isinstance(new_designs, pd.DataFrame):\n",
    "                new_designs_filtered = new_designs.iloc[:, selected_feature_indices]  # Filter columns\n",
    "            else:\n",
    "                new_designs_filtered = new_designs[:, selected_feature_indices]  # Filter columns\n",
    "    \n",
    "            # Filter the scaler parameters to match the selected features\n",
    "            scaler_mean_filtered = np.mean(new_designs_filtered, axis=0)\n",
    "            scaler_std_filtered = np.std(new_designs_filtered, axis=0)\n",
    "\n",
    "            # Standardize new designs using the filtered scaler parameters\n",
    "            epsilon = 1e-8\n",
    "            safe_std = np.where(scaler_std_filtered < epsilon, epsilon, scaler_std_filtered)\n",
    "            new_designs_scaled = (new_designs_filtered - scaler_mean_filtered) / safe_std\n",
    "            if np.any(scaler_std_filtered < epsilon):\n",
    "                print(f\"Warning: {np.sum(scaler_std_filtered < epsilon)} features had near-zero variance\")\n",
    "\n",
    "            # Transform into latent variables\n",
    "            new_designs_latent = new_designs_scaled @ params[\"pls_loadings\"]\n",
    "    \n",
    "            # Compute decision scores and probabilities\n",
    "            decision_scores = new_designs_latent @ params[\"logreg_coefficients\"].T\n",
    "            probabilities = 1 / (1 + np.exp(-decision_scores))\n",
    "\n",
    "            # Store probabilities for this model\n",
    "            rankings[model_name] = probabilities\n",
    "        \n",
    "        # Combine probabilities based on target type\n",
    "        if target_type == \"Truncated\":\n",
    "            # Use OFF_AVG, ON_AVG_Truncated, and ON_OFF_Truncated\n",
    "            # In paper, rankings for \"OFF AVG\" and \"ON AVG Truncated\" were used in Fig. 4\n",
    "            combined_scores = (\n",
    "                #rankings[\"OFF AVG\"]  +\n",
    "                #rankings[\"ON AVG Truncated\"]  +\n",
    "                rankings[\"ON OFF Truncated\"] \n",
    "            )\n",
    "        elif target_type == \"Full\":\n",
    "            # Use OFF_AVG, ON_AVG_Full, and ON_OFF_Full\n",
    "            # In paper, rankings for \"OFF AVG\" and \"ON AVG Full\" were used in Fig. 4\n",
    "            combined_scores = (\n",
    "                #rankings[\"OFF AVG\"]  #+\n",
    "                #rankings[\"ON AVG Full\"]  #+\n",
    "                rankings[\"ON OFF Full\"] \n",
    "            )\n",
    "        else:\n",
    "            raise ValueError(\"Invalid target_type. Choose 'Truncated' or 'Full'.\")\n",
    "\n",
    "        \n",
    "        # Rank designs based on combined scores\n",
    "        ranked_indices = np.argsort(-combined_scores.flatten())  # Negative sign for descending order\n",
    "        final_ranked_scores = combined_scores[ranked_indices]\n",
    "    \n",
    "        # Create a DataFrame for the final ranking\n",
    "        final_ranking = pd.DataFrame({\n",
    "            \"Rank\": range(1, len(ranked_indices) + 1),\n",
    "            \"Combined Score\": final_ranked_scores.flatten(),\n",
    "        })\n",
    "    \n",
    "        return final_ranking, ranked_indices        \n",
    "\n",
    "    def sensor_scoring_VISTA(self, designfile):\n",
    "        \"\"\"\n",
    "        Filters out invalid designs not containing a START codon (AUG), or those that have an in-frame STOP codon.\n",
    "        \n",
    "        Parameters:\n",
    "            mRNA_RR_results_complete (CSV): File containing all computed parameters and riboregulator designs (shape: n_designs x n_features).\n",
    "    \n",
    "        Returns:\n",
    "            {self.design_type}_{self.target_type}_{self.mRNA_name}_all_designs_ranked (CSV): Final ranking of all designs with combined scores.\n",
    "        \"\"\"\n",
    "        design_data = pd.read_csv(\"mRNA_RR_results_complete.csv\")\n",
    "        #design_data.drop(index=0, inplace=True)\n",
    "        \n",
    "        # Extract feature names (excluding the first two columns: 'Target Sequence' and 'Switch Sequence')\n",
    "        feature_names = [col for i, col in enumerate(design_data.columns) if i not in [0, 1]]\n",
    "        \n",
    "        design_spec_data = []\n",
    "        RBS_seq = 'AGAGGAGA'\n",
    "        for _, row in design_data.iterrows():\n",
    "            target_seq = self.dna_to_rna(row['Target Sequence'])\n",
    "            sensor_seq = self.dna_to_rna(row['Switch Sequence'])\n",
    "            index_RBS = sensor_seq.find(RBS_seq)\n",
    "            index_linker = sensor_seq.find(self.hairpin_suffix)\n",
    "            index_AUG = [i for i in range(len(sensor_seq)) if sensor_seq.startswith('AUG', i)]\n",
    "            start_codon_index = 0\n",
    "            for idx in index_AUG:\n",
    "                if idx < index_RBS:\n",
    "                    continue\n",
    "                if (index_linker - idx) % 3 == 0:\n",
    "                    start_codon_index = idx\n",
    "                    break\n",
    "            spec_row = [float(value) for i, value in enumerate(row) if i not in [1, 2]]\n",
    "            if start_codon_index == 0:\n",
    "                print(f'***** {self.current_RNA}: Position {spec_row[0]} of {len(self.mRNA_seq)-self.target_len} ERROR: Suitable start codon not found. Continuing... ***** \\n')\n",
    "                spec_row[0] = 0\n",
    "                design_spec_data.append(spec_row)\n",
    "                continue\n",
    "            aa_seq = self.nt_to_aa(sensor_seq[start_codon_index:])\n",
    "            if '*' in aa_seq:\n",
    "                print(f'***** {self.current_RNA}: Position {spec_row[0]} of {len(self.mRNA_seq)-self.target_len} ERROR: In-Frame STOP codon. Continuing... ***** \\n')\n",
    "                spec_row[0] = 0\n",
    "                design_spec_data.append(spec_row)\n",
    "                continue\n",
    "            design_spec_data.append(spec_row)\n",
    "\n",
    "        design_spec_data = np.array(design_spec_data)  # Convert to NumPy array for easier indexing\n",
    "        \n",
    "        discard_set = []\n",
    "        keep_set = []\n",
    "        # Filter out designs with invalid features--missing start codon or in-frame STOP codon\n",
    "        for c2 in range(design_spec_data.shape[0]):  # Iterate over rows\n",
    "            if design_spec_data[c2, 0] == 0:  # Check first element of each row\n",
    "                discard_set.append(c2)\n",
    "            else:\n",
    "                keep_set.append(c2)\n",
    "        # Keep only valid designs\n",
    "        filtered_designs = design_spec_data[keep_set, 1:]\n",
    "        # Rank the filtered designs\n",
    "        file_name = \"all_trained_model_params.pkl\"\n",
    "        full_path = os.path.join(self.path, file_name) # Use os.path.join() to combine the directory path and file name\n",
    "        final_ranking, ranked_indices = self.rank_new_designs(filtered_designs, self.target_type, full_path)\n",
    "\n",
    "        # Initialize rank and score columns in design_data\n",
    "        design_data[\"Rank\"] = np.nan\n",
    "        design_data[\"Combined Score\"] = np.nan\n",
    "\n",
    "        # Map ranks and scores back to the original design_data using ranked_indices\n",
    "        for i, idx in enumerate(ranked_indices):\n",
    "            original_idx = keep_set[idx]  # Map ranked index back to the original index\n",
    "            design_data.at[design_data.index[original_idx], \"Rank\"] = final_ranking.iloc[i][\"Rank\"]\n",
    "            design_data.at[design_data.index[original_idx], \"Combined Score\"] = final_ranking.iloc[i][\"Combined Score\"]\n",
    "\n",
    "        # Save the ranked designs to a new CSV file\n",
    "        design_data.to_csv(f'{self.design_type}_{self.target_type}_{self.current_RNA}_all_designs_ranked.csv', index=False)    \n",
    "\n",
    "    \n",
    "    def return_design_object(self): \n",
    "        #This function will define the secondary structure of the toehold switch\n",
    "        os.chdir(self.design_type)\n",
    "        for c1 in range(len(self.mRNA_names0)):\n",
    "            self.current_RNA =  self.mRNA_names0.iloc[c1, 0]\n",
    "            os.mkdir(self.current_RNA);\n",
    "            os.chdir(self.current_RNA);\n",
    "            self.mRNA_seq = self.mRNA_sequences0.iloc[c1, 0];\n",
    "            self.T = float(self.temp.iloc[c1].iloc[0]); #in NUPACK 4, temperature will be changed when specifying a physical model\n",
    "            #Define a physical model, as NUPACK 4 analysis and design jobs are run based on a physical model created using the Model class:\n",
    "            self.my_model = Model(material='rna', ensemble='stacking', celsius=self.T,sodium=1.0, magnesium=0.0)\n",
    "\n",
    "            #Initialize the lenghts of each domain in the canonical toehold switch design\n",
    "            self.len_a = 3; self.len_b = 3; self.len_c = 30; self.len_d = 3; self.d_dom = 'AAC'; self.toehold_len = self.len_c;\n",
    "            self.target_len = sum([self.len_a,self.len_b,self.len_c]); self.hairpin_suffix_len = 21; self.output_seq30_len = 30;\n",
    "    \n",
    "            self.output_table = {\"Position\": [], \"Target Sequence\": [], \"Switch Sequence\":[]}\n",
    "            self.output_table[f\"Ascending Stem MFE\"] = [];self.output_table[f\"Descending Stem MFE\"] = [];self.output_table[f\"RBS-GFP Stem MFE\"] = [];self.output_table[f\"RBS-Linker MFE\"] = [];self.output_table[f\"Switch Off MFE\"] = [];self.output_table[f\"Switch Off GFP MFE\"] = [];self.output_table[f\"Switch Off NoTH MFE\"] = [];self.output_table[f\"Hairpin MFE\"] = [];self.output_table[f\"TH+Linker/GFP MFE\"] = [];self.output_table[f\"Switch ON MFE\"] = [];self.output_table[f\"Switch ON-GFP MFE\"] = []\n",
    "            self.output_table[f\"Ascending Stem IED\"] = [];self.output_table[f\"Descending Stem IED\"] = [];self.output_table[f\"RBS-GFP Stem IED\"] = [];self.output_table[f\"RBS-Linker IED\"] = [];self.output_table[f\"Switch Off IED\"] = [];self.output_table[f\"Switch Off GFP IED\"] = [];self.output_table[f\"Switch Off NoTH IED\"] = [];self.output_table[f\"Hairpin IED\"] = [];self.output_table[f\"TH+Linker/GFP IED\"] = [];self.output_table[f\"Switch ON IED\"] = [];self.output_table[f\"Switch ON-GFP IED\"] = []\n",
    "            self.output_table[f\"Ascending Stem NED\"] = [];self.output_table[f\"Descending Stem NED\"] = [];self.output_table[f\"RBS-GFP Stem NED\"] = [];self.output_table[f\"RBS-Linker NED\"] = [];self.output_table[f\"Switch Off NED\"] = [];self.output_table[f\"Switch Off GFP NED\"] = [];self.output_table[f\"Switch Off NoTH NED\"] = [];self.output_table[f\"Hairpin NED\"] = [];self.output_table[f\"TH+Linker/GFP NED\"] = [];self.output_table[f\"Switch ON NED\"] = [];self.output_table[f\"Switch ON-GFP NED\"] = []\n",
    "            self.output_table[f\"First Codon Fraction\"]= []; self.output_table[f\"Second Codon Fraction\"]= []; self.output_table[f\"Average First Two codons\"]= []; self.output_table[f\"Average Target Fraction\"]= [];\n",
    "    \n",
    "            #Here, the flanking lenghts of the 36nt trigger are specified--these are used to compute Ensemble Defect and MFE surrounding the trigger binding site\n",
    "            self.flanking_lengths=[0,10, 25, 50, 100]\n",
    "            for length in self.flanking_lengths:\n",
    "                self.output_table[f\"Trigger (+-{length} nt) Ensemble Defect Ideal\"] = []; self.output_table[f\"Trigger (+-{length} nt) Ensemble Defect Native\"] = []; self.output_table[f\"Trigger (+-{length} nt) MFE\"] = []; \n",
    "    \n",
    "            for n in range(4, 11):\n",
    "                self.output_table[f\"Preceding {n} codons avg\"]= []\n",
    "                self.output_table[f\"Following {n} codons avg\"]= []\n",
    "\n",
    "            #Here, we specify the structure of the toehold switch\n",
    "            self.hairpin_struc_linker_r = 'U%s D%s D3 (U3 D6 U11 U3) U%s U%s'%(self.len_c,(self.len_a+self.len_b),(self.len_d+self.len_a+3),(self.hairpin_suffix_len+self.output_seq30_len))\n",
    "            self.target_struc = 'U%s'%(self.target_len)\n",
    "            self.complex_struc_linker_r = 'D%s + U%s D6 U11 U3 D%s U%s U%s'%((self.toehold_len+self.len_a+self.len_b),(9-(self.len_a+self.len_b)), (3+self.len_a), (self.len_b+self.len_d), (self.hairpin_suffix_len+self.output_seq30_len))\n",
    "            \n",
    "            #The following code was commented out to compute rankings for mCherry in Figure 4. When designing new toehold switches, this is the script that calls all dependent functions defined above \n",
    "            for c2 in range(len(self.mRNA_seq)-self.target_len):\n",
    "                print('%s: Position %s of %s ... \\n'%(self.mRNA_names0.iloc[c1, 0],c2+1, len(self.mRNA_seq)-self.target_len));\n",
    "                target_seq = self.mRNA_seq[c2:c2+self.target_len];\n",
    "                target_seq_comp = self.reverse_complement_seq(target_seq);\n",
    "                toehold = target_seq_comp[:self.toehold_len];\n",
    "                a_dom = target_seq[:self.len_a];\n",
    "                b_dom = target_seq[self.len_a:self.len_a+self.len_b];\n",
    "                a_dom_star = self.reverse_complement_seq(a_dom);\n",
    "                b_dom_star = self.reverse_complement_seq(b_dom);\n",
    "                self.output_seq30 = self.output_sequence.iloc[c1, 0][:30]\n",
    "\n",
    "                hairpin_seq = \"\".join([toehold,b_dom_star,a_dom_star,self.hairpin_seq_top,a_dom,b_dom,self.d_dom,a_dom_star,reverse_complement(self.hairpin_seq_top[-3:]),self.hairpin_suffix,self.output_seq30]);\n",
    "                self.sensor_sequence_r =  self.dna_to_rna(hairpin_seq); self.target_sequence = self.dna_to_rna(target_seq);\n",
    "                self.output_table[\"Position\"].append(c2); self.output_table[\"Target Sequence\"].append(self.target_sequence); self.output_table[\"Switch Sequence\"].append(self.sensor_sequence_r);\n",
    "                self.compute_sensor_parameters(self.sensor_sequence_r)\n",
    "                self.compute_target_parameters(self.target_sequence)\n",
    "                self.compute_codon_fractions(self.mRNA_seq, self.target_sequence, self.codon_usage_dict)\n",
    "            \n",
    "            df = pd.DataFrame(self.output_table)\n",
    "            df.to_csv(\"mRNA_RR_results_complete.csv\", index=False)\n",
    "            ##End commented out section used in Figure 4\n",
    "            #sensor_scoring_VISTA used to score toehold switches based on selected thermodynamic and structural features, as selected by RFE prior to PLS_DA\n",
    "            \n",
    "            self.sensor_scoring_VISTA(\"RNA_RR_results_complete.csv\")\n",
    "            \n",
    "            os.chdir(\"..\")\n",
    "\n",
    "        os.chdir(\"..\")\n",
    "        self.format_final_designs(self.num_designs)\n",
    "\n",
    "    def format_final_designs(self, num_designs):\n",
    "        \"\"\"\n",
    "        Formats the top N designs from the ranked results into the specified structured format.\n",
    "        \n",
    "        Args: num_designs (int): Number of top designs to include in the output file (default 6)\n",
    "                \n",
    "        Returns: writes a CSV file with designs formatted as specified\n",
    "        \"\"\"\n",
    "        # Initialize output lines\n",
    "        output_lines = []\n",
    "\n",
    "        # Create a dictionary mapping mRNA names to sequences\n",
    "        mrna_dict = dict(zip(self.mRNA_names0['Gene Name'], self.mRNA_sequences0['Gene Sequence']))\n",
    "        \n",
    "        # Get the list of RNA directories\n",
    "        rna_dirs = [d for d in os.listdir(self.design_type) if os.path.isdir(os.path.join(self.design_type, d))]\n",
    "    \n",
    "        for rna in rna_dirs:\n",
    "            rna_path = os.path.join(self.design_type, rna)\n",
    "            ranked_file = f\"{self.design_type}_{self.target_type}_{rna}_all_designs_ranked.csv\"\n",
    "            ranked_path = os.path.join(rna_path, ranked_file)\n",
    "            \n",
    "            if os.path.exists(ranked_path):\n",
    "                # Read the ranked designs file\n",
    "                df = pd.read_csv(ranked_path)\n",
    "                \n",
    "                # Filter out invalid designs (those with Rank = NaN)\n",
    "                valid_designs = df[df['Rank'].notna()]\n",
    "                \n",
    "                # Sort by Rank and take top N designs\n",
    "                top_designs = valid_designs.sort_values('Rank').head(num_designs)\n",
    "\n",
    "                # Calculate the average score from the top N combined scores\n",
    "                avg_score = top_designs['Combined Score'].mean()\n",
    "                avg_score_str = f\"{avg_score:.1f}\" if not pd.isna(avg_score) else \"nan\"\n",
    "                \n",
    "                # Get the full target sequence\n",
    "                full_target_seq = mrna_dict.get(rna, \"SEQUENCE_NOT_FOUND\")\n",
    "                \n",
    "                # Add header information\n",
    "                self.name_output = self.output_name.iloc[0].iloc[0]\n",
    "                design_name = f\"{rna}_{self.name_output}_T{int(float(self.T))}\"\n",
    "                output_lines.append(f\"{design_name} target:,,,\")\n",
    "                output_lines.append(\"\")\n",
    "                output_lines.append(f\"{rna} RNA sequence,{full_target_seq},,,\")\n",
    "                output_lines.append(f\"Target region {rna}: Average design score (top {num_designs} devices) =, {avg_score_str},,,\")\n",
    "                output_lines.append(\"\")\n",
    "                output_lines.append(\"Sensor name,Index,Sensor RNA sequence,Target RNA subsequence\")\n",
    "                \n",
    "                # Add each sensor design\n",
    "                for i, (_, row) in enumerate(top_designs.iterrows(), 1):\n",
    "                    sensor_name = f\"VISTA_{design_name}_{chr(64+i)}\"  # A, B, C, etc.\n",
    "                    sensor_index = row['Position']\n",
    "                    sensor_seq = self.rna_to_dna(row['Switch Sequence'])\n",
    "                    target_subseq = self.rna_to_dna(row['Target Sequence'])\n",
    "                    \n",
    "                    output_lines.append(f\"{sensor_name},{sensor_index},{sensor_seq},{target_subseq}\")\n",
    "                output_lines.append(\"\")\n",
    "                output_lines.append(\"\")\n",
    "\n",
    "\n",
    "        # Write to CSV\n",
    "        output_path = os.path.join(self.design_type, \"VISTA_final_design_info.csv\")\n",
    "        \n",
    "        # Use newline='' to prevent extra blank lines in Windows\n",
    "        with open(output_path, 'w', newline='') as f:\n",
    "            f.write('\\n'.join(output_lines))\n",
    "        \n",
    "        print(f\"Successfully saved formatted designs to {output_path}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d9b877e7-979b-4d88-8c4e-bd448ee3e7ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-08-12 13:11:07.241 python[61988:2887316] +[CATransaction synchronize] called within transaction\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ryhB_sRNA: Position 1 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 2 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 3 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 4 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 5 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 6 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 7 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 8 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 9 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 10 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 11 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 12 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 13 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 14 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 15 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 16 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 17 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 18 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 19 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 20 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 21 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 22 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 23 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 24 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 25 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 26 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 27 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 28 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 29 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 30 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 31 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 32 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 33 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 34 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 35 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 36 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 37 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 38 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 39 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 40 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 41 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 42 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 43 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 44 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 45 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 46 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 47 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 48 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 49 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 50 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 51 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 52 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 53 of 54 ... \n",
      "\n",
      "ryhB_sRNA: Position 54 of 54 ... \n",
      "\n",
      "***** ryhB_sRNA: Position 4.0 of 54 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "***** ryhB_sRNA: Position 25.0 of 54 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "***** ryhB_sRNA: Position 28.0 of 54 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "***** ryhB_sRNA: Position 45.0 of 54 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "Warning: 2 features had near-zero variance\n",
      "Warning: 1 features had near-zero variance\n",
      "Warning: 1 features had near-zero variance\n",
      "spot42_sRNA: Position 1 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 2 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 3 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 4 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 5 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 6 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 7 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 8 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 9 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 10 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 11 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 12 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 13 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 14 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 15 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 16 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 17 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 18 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 19 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 20 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 21 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 22 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 23 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 24 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 25 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 26 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 27 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 28 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 29 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 30 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 31 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 32 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 33 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 34 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 35 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 36 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 37 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 38 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 39 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 40 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 41 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 42 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 43 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 44 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 45 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 46 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 47 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 48 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 49 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 50 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 51 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 52 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 53 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 54 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 55 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 56 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 57 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 58 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 59 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 60 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 61 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 62 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 63 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 64 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 65 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 66 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 67 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 68 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 69 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 70 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 71 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 72 of 73 ... \n",
      "\n",
      "spot42_sRNA: Position 73 of 73 ... \n",
      "\n",
      "***** spot42_sRNA: Position 1.0 of 73 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "***** spot42_sRNA: Position 11.0 of 73 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "***** spot42_sRNA: Position 14.0 of 73 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "***** spot42_sRNA: Position 23.0 of 73 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "***** spot42_sRNA: Position 30.0 of 73 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "***** spot42_sRNA: Position 39.0 of 73 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "***** spot42_sRNA: Position 44.0 of 73 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "***** spot42_sRNA: Position 46.0 of 73 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "***** spot42_sRNA: Position 49.0 of 73 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "***** spot42_sRNA: Position 60.0 of 73 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "***** spot42_sRNA: Position 63.0 of 73 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "***** spot42_sRNA: Position 69.0 of 73 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "***** spot42_sRNA: Position 71.0 of 73 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "***** spot42_sRNA: Position 72.0 of 73 ERROR: In-Frame STOP codon. Continuing... ***** \n",
      "\n",
      "Warning: 2 features had near-zero variance\n",
      "Warning: 1 features had near-zero variance\n",
      "Warning: 1 features had near-zero variance\n",
      "Successfully saved formatted designs to Toehold_VISTA/VISTA_final_design_info.csv\n"
     ]
    }
   ],
   "source": [
    "# Here, you will call the design class and execute the design process. \n",
    "\n",
    "# First, update source_seq_list with your required sequences, output/reporter, and temperature. \n",
    "\n",
    "# Under target_type please select whether you would like to optimize for targeting a \"Truncated\" or \"Full\" length RNA transcript.\n",
    "\n",
    "# Num_designs allows you to select the number of toehold switches you would like the code to return. \n",
    "\n",
    "# The code will ask you to select a directory. When the window launches, please select the working directory with the necessary files...\n",
    "# ... for the code to run. These should all be in the toehold-VISTA folder. \n",
    "\n",
    "\n",
    "toehold_VISTA = toehold_switch_design_object(\n",
    "    source_seq_list='source_seq_list.csv', # Update the source_seq_list.csv in the toehold_VISTA folder to include Gene Name, Gene Sequence, Temperature, Output Name, and Output Sequence\n",
    "    ecoli_codon_usage_table='ecoli_codon_usage_table.csv', # This file should not change--maintain this file in the toehold_VISTA folder for the code to run\n",
    "    target_type='Full', # Select which toehold switch targeting optimization you desire-- select from \"Full\" or \"Truncated\"\n",
    "    num_designs=12 # Select the number of designs you wish the code to select. \n",
    ")\n",
    "toehold_VISTA.return_design_object()  # Execute the design process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93d7b11a-9865-4e8d-8af0-4c029f04d34e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dadfaf29-554a-40c6-a612-ffb3fcfd04e6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
